File: Fibonacci Sequence/Fibonacci Sequence.rs

// Time Complexity: O(n), where n is the length of the sequence to generate.
// Space Complexity: O(n), for storing the output sequence of length n.
fn fibonacci_sequence(start: [i64; 2], length: usize) -> Vec<i64> {
    if length == 0 {
        return vec![];
    } else if length == 1 {
        return vec![start[0]];
    } else if length == 2 {
        return vec![start[0], start[1]];
    }

    let mut seq = vec![start[0], start[1]];
    while seq.len() < length {
        let n = seq.len();
        let next = seq[n - 1] + seq[n - 2];
        seq.push(next);
    }
    seq
}

fn main() {
    println!("{:?}", fibonacci_sequence([0, 1], 10));
    println!("{:?}", fibonacci_sequence([123456789, 987654321], 5));
}

File: Fibonacci Sequence/Fibonacci Sequence.c
#include <stdio.h>

void fibonacci_sequence(long long start1, long long start2, int length) {
    if (length == 0) return;
    if (length == 1) {
        printf("%lld\n", start1);
        return;
    }
    if (length == 2) {
        printf("%lld %lld\n", start1, start2);
        return;
    }

    long long a = start1, b = start2;
    printf("%lld %lld", a, b);
    for (int i = 2; i < length; i++) {
        long long next = a + b;
        printf(" %lld", next);
        a = b;
        b = next;
    }
    printf("\n");
}

int main() {
    fibonacci_sequence(0, 1, 10);
    fibonacci_sequence(123456789LL, 987654321LL, 5);
    return 0;
}
